File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int sum=1,c=0;
        for(int i=0;i<nums.size();i++){
            if(nums[i]!=0) sum=sum*nums[i];
            else c+=1;
        }
        vector<int>ans;
        for(int i=0;i<nums.size();i++){
            if(c>1) ans.push_back(0);
            else if(nums[i]==0) ans.push_back(sum);
            else if(c>0) ans.push_back(0);
            else ans.push_back(sum/nums[i]);
        }
        return ans;
    }
};

File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int zeros=0,mul1=1,mul2=1;
        int[] ans = new int[nums.length];
        for(int i=0;i<nums.length;i++){
            if(nums[i]==0){
                zeros++;
                if(zeros==2) break;
                mul2=mul1;
                mul1=1;
            }
            else{
                mul1=mul1*nums[i];
            }
        }
        if(zeros==2) return ans;
        else if(zeros==1){
            for(int i=0;i<nums.length;i++){
                if(nums[i]==0){
                    ans[i]=mul1*mul2;
                    break;
                }
            }
            return ans;
        }
        for(int i=0;i<nums.length;i++){
            ans[i]=mul1/nums[i];
        }
        return ans;
    }
}